### ISSUE TITLE
Dimensional mismatch in the evaluation step of transfer learning

### ISSUE THREAD

**Issue by nimning** [Microsoft/CNTK#3348]
The evaluation is defined as the following

```
def eval_single_image(loaded_model, image_path, image_dims):
    # load and format image (resize, RGB -> BGR, CHW -> HWC)
    try:
        img = Image.open(image_path)
        
        #convert to RGB
        if image_path.endswith("png"):
            temp = Image.new("RGB", img.size, (255, 255, 255))
            temp.paste(img, img)
            img = temp

        resized = img.resize((image_dims[2], image_dims[1]), Image.ANTIALIAS)
        bgr_image = np.asarray(resized, dtype=np.float32)[..., [2, 1, 0]]
        hwc_format = np.ascontiguousarray(np.rollaxis(bgr_image, 2))

        # compute model output
        arguments = {loaded_model.arguments[0]: [hwc_format]}
        output = loaded_model.eval(arguments)
...       
```

The `resized` has the same dimension as the input image` (width, height, channel)`. So, the dimension of `bgr_image` is `(width, height,flip(channel))` and `hwc_format` is actual `(flip(channel), width, height)`. However, the dimension of `loaded_model.arguments[0]` input is `(channel, height, width)`. The dimension does not match. This is quite confusing.

Also, I do not see why it is necessary to flip the color channel to convert BGR to RGB. We did not do similar transformation in training step. Why it is necessary in test step?

**Comment by KeDengMS** [comment]
It has some historical background and the terms are confusing. First, the comments on CHW->HWC is incorrect. It's actually HWC->CHW. You may find some explanation of NCHW vs. NHWC format [here](https://www.tensorflow.org/performance/performance_guide). CNTK uses NCHW as input, and that why `loaded_model.arguments[0]` is `(channel, height, weight)`.

As for RGB->BGR, it has something to do with endianness. For 8-bit per-channel image, the byte order of RGB in memory would become BGR when the all channels of pixel are fetched as a single int32 (with 8-bit padding or alpha) on little-endian CPUs. The best way to make sure you are loading channels correctly is to create an image with pure R/G/B color, and check the data feed into network to make sure the values are expected.

**Comment by nimning** [comment]
I see. Thanks!

**Comment by nimning** [comment]
Hi KeDengMS

I understand now that CNTK uses column major and the need to flip the channel. However, the `np.rollaxis` only flips the channel dimension. So, the the dimension of ` hwc_format ` is actual `(flip(channel), width, height)`.  The `loaded_model.arguments[0]` is `(channel, height, weight)`. The `height` and `width `still mismatch between the argument and value. 

It seems an additional transformation needs to be applied on `hwc_format` to make the dimension as `(flip(channel), height, width)`
